backend/metadata_writer/src/metadata_writer.py

```diff
@@ -0,0 +1,333 @@
+sleep_time = 0.1
+
+
+def patch_pod_metadata(
+    namespace: str,
+    pod_name: str,
+    patch: dict,
+    k8s_api: kubernetes.client.CoreV1Api = None,
+):
+    k8s_api = k8s_api or kubernetes.client.CoreV1Api()
+    patch = {
+        'metadata': patch
+    }
+    for retry in range(patch_retries):
+        try:
+            pod = k8s_api.patch_namespaced_pod(
+                name=pod_name,
+                namespace=namespace,
+                body=patch,
+            )
+            return pod
+        except Exception as e:
+            print(e)
+            sleep(sleep_time)
+
+
+#Connecting to MetadataDB
+mlmd_store = connect_to_mlmd()
+print("Connected to the metadata store")
+
+
+ARGO_OUTPUTS_ANNOTATION_KEY = 'workflows.argoproj.io/outputs'
+ARGO_TEMPLATE_ANNOTATION_KEY = 'workflows.argoproj.io/template'
+KFP_COMPONENT_SPEC_ANNOTATION_KEY = 'pipelines.kubeflow.org/component_spec'
+METADATA_EXECUTION_ID_LABEL_KEY = 'pipelines.kubeflow.org/metadata_execution_id'
+METADATA_CONTEXT_ID_LABEL_KEY = 'pipelines.kubeflow.org/metadata_context_id'
+METADATA_ARTIFACT_IDS_ANNOTATION_KEY = 'pipelines.kubeflow.org/metadata_artifact_ids'
+METADATA_INPUT_ARTIFACT_IDS_ANNOTATION_KEY = 'pipelines.kubeflow.org/metadata_input_artifact_ids'
+METADATA_OUTPUT_ARTIFACT_IDS_ANNOTATION_KEY = 'pipelines.kubeflow.org/metadata_output_artifact_ids'
+
+ARGO_WORKFLOW_LABEL_KEY = 'workflows.argoproj.io/workflow'
+ARGO_COMPLETED_LABEL_KEY = 'workflows.argoproj.io/completed'
+METADATA_WRITTEN_LABEL_KEY = 'pipelines.kubeflow.org/metadata_written'
+
+
+def output_name_to_argo(name: str) -> str:
+    import re
+    return re.sub('-+', '-', re.sub('[^-0-9a-z]+', '-', name.lower())).strip('-')
+
+
+def argo_artifact_to_uri(artifact: dict) -> str:
+    if 's3' in artifact:
+        s3_artifact = artifact['s3']
+        return 'minio://{bucket}/{key}'.format(
+            bucket=s3_artifact.get('bucket', ''),
+            key=s3_artifact.get('key', ''),
+        )
+    elif 'raw' in artifact:
+        return None
+    else:
+        return None
+
+
+def is_tfx_pod(pod) -> bool:
+    main_containers = [container for container in pod.spec.containers if container.name == 'main']
+    if len(main_containers) != 1:
+        return False
+    main_container = main_containers[0]
+    return main_container.command and main_container.command[-1].endswith('tfx/orchestration/kubeflow/container_entrypoint.py')
+
+
+# Caches (not expected to be persistent)
+# These caches are only used to prevent race conditions. Race conditions happen because the writer can see multiple versions of K8s object before the applied labels show up.
+# They are expected to be lost when restarting the service.
+# The operation of the Metadata Writer remains correct even if it's getting restarted frequently. (Kubernetes only sends the latest version of resource for new watchers.)
+# Technically, we could remove the objects from cache as soon as we see that our labels have been applied successfully.
```

Comment:
I agree. In the future I should find some LRU collection that removes the old items.

Initially this won't be a problem since the amount of data is still very small and the caches will be purged if the service restarts.